File: core/src/main/java/me/liam/support/ExtraTransaction.java
package me.liam.support;

import android.os.Bundle;
import android.os.Handler;
import android.view.animation.AnimationUtils;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import java.util.List;

import me.liam.anim.FragmentAnimation;
import me.liam.fragmentation.R;
import me.liam.helper.FragmentUtils;
import me.liam.queue.Action;
import me.liam.queue.ActionQueue;

public abstract class ExtraTransaction {

    public abstract ExtraTransaction setCustomerAnimations(int enterAnim,int exitAnim);

    public abstract ExtraTransaction setCustomerAnimations(int enterAnim,int exitAnim,int popEnterAnim,int popExitAnim);

    public abstract ExtraTransaction dontDisplaySelfPopAnim(boolean dontDisplaySelfPopAnim);

    public abstract ExtraTransaction setResult(int resultCode, Bundle data);

    public abstract ExtraTransaction displayEnterAnim(boolean displayEnterAnim);

    public abstract ExtraTransaction setTag(String tag);

    public abstract ExtraTransaction addBackStack(boolean addBackStack);

    public abstract ExtraTransaction runOnExecute(Runnable run);

    public abstract void loadRootFragment(int containerId, SupportFragment to);

    public abstract void show(SupportFragment... show);

    public abstract void hide(SupportFragment... hide);

    public abstract void start(SupportFragment to);

    public abstract void startWithPop(SupportFragment to);

    public abstract void startWithPopTo(SupportFragment to, Class popToCls);

    public abstract void startWithPopTo(SupportFragment to, Class popToCls,boolean includeTarget);

    public abstract void startForResult(SupportFragment to,int requestCode);

    public abstract void pop();

    public abstract void popChild();

    public abstract void popTo(Class popToCls);

    public abstract void popTo(Class popToCls,boolean includeTarget);

    public abstract void popChildTo(Class popToCls);

    public abstract void popChildTo(Class popToCls,boolean includeTarget);

    public abstract void remove(SupportFragment remove);

    public abstract void remove(SupportFragment remove, boolean anim);

    public abstract void remove(SupportFragment... remove);

    static class ExtraTransactionImpl extends ExtraTransaction {

        private SupportFragment from;

        private TransactionRecord record;

        private ActionQueue actionQueue;

        public ExtraTransactionImpl(SupportFragment from, Handler handler){
            this.from = from;
            record = new TransactionRecord();
            actionQueue = new ActionQueue(handler);
        }

        @Override
        public ExtraTransaction setCustomerAnimations(int enterAnim, int exitAnim) {
            record.fragmentAnimation = new FragmentAnimation(enterAnim,exitAnim, R.anim.anim_empty,R.anim.anim_empty);
            return this;
        }

        @Override
        public ExtraTransaction setCustomerAnimations(int enterAnim, int exitAnim, int popEnterAnim, int popExitAnim) {
            record.fragmentAnimation = new FragmentAnimation(enterAnim,exitAnim, popEnterAnim, popExitAnim);
            return this;
        }

        @Override
        public ExtraTransaction dontDisplaySelfPopAnim(boolean dontDisplaySelfPopAnim) {
            record.dontDisplaySelfPopAnim = dontDisplaySelfPopAnim;
            return this;
        }

        @Override
        public ExtraTransaction setResult(int resultCode, Bundle data) {
            record.resultCode = resultCode;
            record.resultData = data;
            setResult(from,resultCode,data);
            return this;
        }

        @Override
        public ExtraTransaction displayEnterAnim(boolean displayEnterAnim) {
            record.displayEnterAnim = displayEnterAnim;
            return this;
        }

        @Override
        public ExtraTransaction setTag(String tag) {
            record.tag = tag;
            return this;
        }

        @Override
        public ExtraTransaction addBackStack(boolean addBackStack) {
            record.addBackStack = addBackStack;
            return this;
        }


        @Override
        public ExtraTransaction runOnExecute(Runnable run) {
            record.runOnExecute = run;
            return this;
        }

        @Override
        public void loadRootFragment(final int containerId, final SupportFragment to) {
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    bindFragmentOptions(to,containerId,record);
                    to.setFragmentAnimation(record.fragmentAnimation);
                    FragmentTransaction ft = from.getChildFragmentManager().beginTransaction();
                    ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                    ft.add(containerId,to,record.tag);
                    supportCommit(ft,record.runOnExecute);
                    return 0;
                }
            });
        }

        @Override
        public void show(final SupportFragment... show) {
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                    for (SupportFragment f : show){
                        ft.show(f);
                    }
                    supportCommit(ft,record.runOnExecute);
                    return 0;
                }
            });
        }

        @Override
        public void hide(final SupportFragment... hide) {
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                    for (SupportFragment f : hide){
                        ft.hide(f);
                    }
                    supportCommit(ft,record.runOnExecute);
                    return 0;
                }
            });
        }

        @Override
        public void start(final SupportFragment to) {
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    bindFragmentOptions(to,from.getContainerId(),record);
                    to.setFragmentAnimation(record.fragmentAnimation);
                    FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                    ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                    ft.add(from.getContainerId(),to,record.tag);
                    supportCommit(ft,record.runOnExecute);
                    return 0;
                }
            });
        }

        @Override
        public void startWithPop(final SupportFragment to) {
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    bindFragmentOptions(to,from.getContainerId(),record);
                    to.setFragmentAnimation(record.fragmentAnimation);
                    FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                    ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                    ft.add(from.getContainerId(),to,record.tag);
                    supportCommit(ft,record.runOnExecute);
                    to.setCallBack(new SupportFragmentCallBack(){
                        @Override
                        public void onEnterAnimEnd() {
                            FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                            ft.remove(from);
                            supportCommit(ft);
                        }
                    });
                    return 0;
                }
            });
        }

        @Override
        public void startWithPopTo(SupportFragment to, Class popToCls) {
            startWithPopTo(to,popToCls,true);
        }

        @Override
        public void startWithPopTo(final SupportFragment to, Class popToCls, boolean includeTarget) {
            record.popToCls = popToCls;
            record.includeTarget = includeTarget;
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    bindFragmentOptions(to,from.getContainerId(),record);
                    to.setFragmentAnimation(record.fragmentAnimation);
                    FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                    ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
                    ft.add(from.getContainerId(),to,record.tag);
                    supportCommit(ft,record.runOnExecute);
                    to.setCallBack(new SupportFragmentCallBack(){
                        @Override
                        public void onEnterAnimEnd() {
                            popTo(from.getFragmentManager(),record.popToCls,record.includeTarget,null);
                        }
                    });
                    return 0;
                }
            });
        }

        @Override
        public void startForResult(SupportFragment to, int requestCode) {
            record.requestCode = requestCode;
            setStartForResult(from,to,requestCode);
            start(to);
        }

        @Override
        public void pop() {
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    SupportFragment remove = FragmentUtils.getLastFragment(from.getFragmentManager());
                    if (remove == null) return 0;
                    long duration = AnimationUtils.loadAnimation(remove.getContext(),remove.getFragmentAnimation().getExitAnimId()).getDuration();
                    FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                    ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE);
                    ft.remove(remove);
                    supportCommit(ft,record.runOnExecute);
                    return duration;
                }

                @Override
                public int actionType() {
                    return Action.TYPE_POP;
                }
            });
        }

        @Override
        public void popChild() {
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    SupportFragment remove = FragmentUtils.getLastFragment(from.getChildFragmentManager());
                    if (remove == null) return 0;
                    long duration = AnimationUtils.loadAnimation(remove.getContext(),remove.getFragmentAnimation().getExitAnimId()).getDuration();
                    FragmentTransaction ft = from.getChildFragmentManager().beginTransaction();
                    ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE);
                    ft.remove(remove);
                    supportCommit(ft,record.runOnExecute);
                    return duration;
                }

                @Override
                public int actionType() {
                    return Action.TYPE_POP;
                }
            });
        }

        @Override
        public void popTo(Class popToCls) {
            record.popToCls = popToCls;
            popTo(record.popToCls,true);
        }

        @Override
        public void popTo(Class popToCls, boolean includeTarget) {
            record.popToCls = popToCls;
            record.includeTarget = includeTarget;
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    popTo(from.getFragmentManager(),record.popToCls,record.includeTarget,record.runOnExecute);
                    return 0;
                }

                @Override
                public int actionType() {
                    return Action.TYPE_POP;
                }
            });
        }

        @Override
        public void popChildTo(Class popToCls) {
            record.popToCls = popToCls;
            popChildTo(record.popToCls,true);
        }

        @Override
        public void popChildTo(Class popToCls, boolean includeTarget) {
            record.popToCls = popToCls;
            record.includeTarget = includeTarget;
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    popTo(from.getChildFragmentManager(),record.popToCls,record.includeTarget,record.runOnExecute);
                    return 0;
                }

                @Override
                public int actionType() {
                    return Action.TYPE_POP;
                }
            });
        }

        @Override
        public void remove(SupportFragment remove) {
            record.remove = remove;
            remove(remove,true);
        }

        @Override
        public void remove(final SupportFragment remove, final boolean anim) {
            record.remove = remove;
            record.removeAnim = anim;
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    if (from == null || from.getFragmentManager() == null) return 0;
                    FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                    if (anim){
                        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE);
                    }
                    ft.remove(remove);
                    supportCommit(ft,record.runOnExecute);
                    return 0;
                }

                @Override
                public int actionType() {
                    return Action.TYPE_POP;
                }
            });
        }

        @Override
        public void remove(final SupportFragment... remove) {
            actionQueue.enqueue(new Action() {
                @Override
                public long run() {
                    if (from == null || from.getFragmentManager() == null) return 0;
                    FragmentTransaction ft = from.getFragmentManager().beginTransaction();
                    for (SupportFragment f : remove){
                        ft.remove(f);
                    }
                    supportCommit(ft,record.runOnExecute);
                    return 0;
                }

                @Override
                public int actionType() {
                    return Action.TYPE_POP;
                }
            });
        }

        Bundle getArguments(SupportFragment target){
            if (target.getArguments() == null){
                target.setArguments(new Bundle());
            }
            return target.getArguments();
        }

        void bindFragmentOptions(SupportFragment target, int containerId,TransactionRecord record) {
            Bundle args = getArguments(target);
            args.putInt(SupportTransaction.FRAGMENTATION_CONTAINER_ID, containerId);
            args.putString(SupportTransaction.FRAGMENTATION_TAG, record.tag);
            args.putString(SupportTransaction.FRAGMENTATION_SIMPLE_NAME, target.getClass().getSimpleName());
            args.putString(SupportTransaction.FRAGMENTATION_FULL_NAME, target.getClass().getName());
            args.putBoolean(SupportTransaction.FRAGMENTATION_PLAY_ENTER_ANIM, record.displayEnterAnim);
            args.putBoolean(SupportTransaction.FRAGMENTATION_INIT_LIST, false);
            args.putBoolean(SupportTransaction.FRAGMENTATION_SAVED_INSTANCE, false);
            args.putBoolean(SupportTransaction.FRAGMENTATION_DONT_DISPLAY_SELF_POP_ANIM,record.dontDisplaySelfPopAnim);
        }

        void supportCommit(FragmentTransaction ft) {
            supportCommit(ft,null);
        }

        void supportCommit(FragmentTransaction ft,Runnable runnable) {
            if (runnable != null){
                ft.runOnCommit(runnable);
            }
            ft.commitAllowingStateLoss();
        }

        void setStartForResult(SupportFragment from, SupportFragment to, int requestCode){
            Bundle formBundle = getArguments(from);
            formBundle.putInt(SupportTransaction.FRAGMENTATION_REQUEST_CODE,requestCode);
            Bundle toBundle = getArguments(to);
            toBundle.putInt(SupportTransaction.FRAGMENTATION_FROM_REQUEST_CODE,requestCode);
        }

        void setResult(SupportFragment target, int resultCode, Bundle data){
            Bundle bundle = getArguments(target);
            bundle.putInt(SupportTransaction.FRAGMENTATION_RESULT_CODE,resultCode);
            bundle.putBundle(SupportTransaction.FRAGMENTATION_RESULT_DATA,data);
        }

        void popTo(FragmentManager fm,Class cls,boolean includeTarget,Runnable run){
            SupportFragment remove = FragmentUtils.getLastFragment(fm);
            SupportFragment target = FragmentUtils.findFragmentByClass(fm,cls);
            if (remove == null || target == null) return;
            int targetIndex = fm.getFragments().indexOf(target);
            int removeIndex = fm.getFragments().indexOf(remove);
            FragmentTransaction ft = fm.beginTransaction();
            ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE);
            List<Fragment> removeList = fm.getFragments().subList(targetIndex,removeIndex);
            if (!includeTarget){
                removeList.remove(target);
            }
            for (Fragment f : removeList){
                if (f instanceof SupportFragment){
                    ft.remove(f);
                }
            }
            ft.remove(remove);
            supportCommit(ft,run);
        }
    }
}

File: core/src/main/java/me/liam/fragmentation/support/ExtraTransaction.kt
package me.liam.fragmentation.support

import android.os.Bundle
import android.os.Handler
import android.view.animation.AnimationUtils
import androidx.annotation.IdRes

import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction

import me.liam.fragmentation.anim.FragmentAnimation
import me.liam.fragmentation.R
import me.liam.fragmentation.helper.findFragment
import me.liam.fragmentation.helper.getLastFragment
import me.liam.fragmentation.helper.supportCommit
import me.liam.fragmentation.queue.Action
import me.liam.fragmentation.queue.ActionQueue

abstract class ExtraTransaction {

    abstract fun setCustomerAnimations(enterAnim: Int, exitAnim: Int, popEnterAnim: Int? = null, popExitAnim: Int? = null): ExtraTransaction

    abstract fun dontDisplaySelfPopAnim(dontDisplaySelfPopAnim: Boolean): ExtraTransaction

    abstract fun setResult(resultCode: Int, data: Bundle): ExtraTransaction

    abstract fun displayEnterAnim(displayEnterAnim: Boolean): ExtraTransaction

    abstract fun setTag(tag: String): ExtraTransaction

    abstract fun addBackStack(addBackStack: Boolean): ExtraTransaction

    abstract fun runOnExecute(run: Runnable): ExtraTransaction

    abstract fun loadRootFragment(@IdRes containerId: Int, to: SupportFragment)

    abstract fun show(vararg show: SupportFragment)

    abstract fun hide(vararg hide: SupportFragment)

    abstract fun start(to: SupportFragment)

    abstract fun startWithPop(to: SupportFragment)

    abstract fun startWithPopTo(to: SupportFragment, popToCls: Class<*>, includeTarget: Boolean = true)

    abstract fun startForResult(to: SupportFragment, requestCode: Int)

    abstract fun pop()

    abstract fun popChild()

    abstract fun popTo(popToCls: Class<*>, includeTarget: Boolean = true)

    abstract fun popChildTo(popToCls: Class<*>, includeTarget: Boolean = true)

    abstract fun remove(remove: SupportFragment, anim: Boolean = true)

    abstract fun remove(vararg remove: SupportFragment)

    internal class ExtraTransactionImpl(private val from: SupportFragment, handler: Handler) : ExtraTransaction() {

        private val record = TransactionRecord()
        private val actionQueue = ActionQueue(handler)

        override fun setCustomerAnimations(enterAnim: Int, exitAnim: Int, popEnterAnim: Int?, popExitAnim: Int?): ExtraTransaction {
            record.fragmentAnimation = FragmentAnimation(enterAnim, exitAnim, popEnterAnim
                    ?: R.anim.anim_empty, popExitAnim ?: R.anim.anim_empty)
            return this
        }

        override fun dontDisplaySelfPopAnim(dontDisplaySelfPopAnim: Boolean): ExtraTransaction {
            record.dontDisplaySelfPopAnim = dontDisplaySelfPopAnim
            return this
        }

        override fun setResult(resultCode: Int, data: Bundle): ExtraTransaction {
            record.resultCode = resultCode
            record.resultData = data
            setResult(from, resultCode, data)
            return this
        }

        override fun displayEnterAnim(displayEnterAnim: Boolean): ExtraTransaction {
            record.displayEnterAnim = displayEnterAnim
            return this
        }

        override fun setTag(tag: String): ExtraTransaction {
            record.tag = tag
            return this
        }

        override fun addBackStack(addBackStack: Boolean): ExtraTransaction {
            record.addBackStack = addBackStack
            return this
        }


        override fun runOnExecute(run: Runnable): ExtraTransaction {
            record.runOnExecute = run
            return this
        }

        override fun loadRootFragment(containerId: Int, to: SupportFragment) {
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    bindFragmentOptions(to, containerId, record)
                    to.fragmentAnimation = record.fragmentAnimation

                    from.childFragmentManager.beginTransaction().apply {
                        setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                        add(containerId, to, record.tag)
                    }.supportCommit(record.runOnExecute)
                    return 0
                }
            })
        }

        override fun show(vararg show: SupportFragment) {
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    from.fragmentManager?.beginTransaction()?.apply {
                        for (f in show) show(f)
                    }?.supportCommit(record.runOnExecute)
                    return 0
                }
            })
        }

        override fun hide(vararg hide: SupportFragment) {
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    from.fragmentManager?.beginTransaction()?.apply {
                        for (f in hide) hide(f)
                    }?.supportCommit(record.runOnExecute)
                    return 0
                }
            })
        }

        override fun start(to: SupportFragment) {
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    bindFragmentOptions(to, from.containerId, record)
                    to.fragmentAnimation = record.fragmentAnimation

                    from.fragmentManager?.beginTransaction()?.apply {
                        setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                        add(from.containerId, to, record.tag)
                    }?.supportCommit(record.runOnExecute)
                    return 0
                }
            })
        }

        override fun startWithPop(to: SupportFragment) {
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    bindFragmentOptions(to, from.containerId, record)
                    to.fragmentAnimation = record.fragmentAnimation

                    from.fragmentManager?.beginTransaction()?.apply {
                        setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                        add(from.containerId, to, record.tag)
                    }?.supportCommit(record.runOnExecute)

                    to.setAnimEndListener {
                        from.fragmentManager?.beginTransaction()?.apply {
                            remove(from);supportCommit()
                        }
                    }
                    return 0
                }
            })
        }

        override fun startWithPopTo(to: SupportFragment, popToCls: Class<*>, includeTarget: Boolean) {
            record.popToCls = popToCls
            record.includeTarget = includeTarget
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    bindFragmentOptions(to, from.containerId, record)
                    to.fragmentAnimation = record.fragmentAnimation

                    from.fragmentManager?.beginTransaction()?.apply {
                        setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                        add(from.containerId, to, record.tag)
                    }?.supportCommit(record.runOnExecute)

                    to.setAnimEndListener {
                        popTo(from.fragmentManager, record.popToCls, record.includeTarget, null)
                    }
                    return 0
                }
            })
        }

        override fun startForResult(to: SupportFragment, requestCode: Int) {
            record.requestCode = requestCode
            setStartForResult(from, to, requestCode)
            start(to)
        }

        override fun pop() {
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    val remove = from.fragmentManager?.getLastFragment() ?: return 0
                    val duration = remove.fragmentAnimation?.exitAnimId?.let { AnimationUtils.loadAnimation(remove.context, it).duration }

                    from.fragmentManager?.beginTransaction()?.apply {
                        setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                        remove(remove)
                    }?.supportCommit(record.runOnExecute)
                    return duration ?: 0
                }

                override fun actionType(): Int {
                    return Action.TYPE_POP
                }
            })
        }

        override fun popChild() {
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    val remove = from.childFragmentManager.getLastFragment()
                            ?: return 0
                    val duration = remove.fragmentAnimation?.exitAnimId?.let { AnimationUtils.loadAnimation(remove.context, it).duration }

                    from.childFragmentManager.beginTransaction().apply {
                        setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                        remove(remove)
                    }.supportCommit(record.runOnExecute)
                    return duration ?: 0
                }

                override fun actionType(): Int {
                    return Action.TYPE_POP
                }
            })
        }

        override fun popTo(popToCls: Class<*>, includeTarget: Boolean) {
            record.popToCls = popToCls
            record.includeTarget = includeTarget

            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    popTo(from.fragmentManager, record.popToCls, record.includeTarget, record.runOnExecute)
                    return 0
                }

                override fun actionType(): Int {
                    return Action.TYPE_POP
                }
            })
        }

        override fun popChildTo(popToCls: Class<*>, includeTarget: Boolean) {
            record.popToCls = popToCls
            record.includeTarget = includeTarget

            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    popTo(from.childFragmentManager, record.popToCls, record.includeTarget, record.runOnExecute)
                    return 0
                }

                override fun actionType(): Int {
                    return Action.TYPE_POP
                }
            })
        }

        override fun remove(remove: SupportFragment, anim: Boolean) {
            record.remove = remove
            record.removeAnim = anim

            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    from.fragmentManager?.beginTransaction()?.apply {
                        if (anim) setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                        remove(remove)
                    }?.supportCommit(record.runOnExecute)
                    return 0
                }

                override fun actionType(): Int {
                    return Action.TYPE_POP
                }
            })
        }

        override fun remove(vararg remove: SupportFragment) {
            actionQueue.enqueue(object : Action() {
                override fun run(): Long {
                    from.fragmentManager?.beginTransaction()?.apply {
                        for (f in remove) remove(f)
                    }?.supportCommit(record.runOnExecute)
                    return 0
                }

                override fun actionType(): Int {
                    return Action.TYPE_POP
                }
            })
        }

        private fun getArguments(target: SupportFragment?): Bundle {
            return target?.arguments ?: Bundle().also { target?.arguments = it }
        }

        private fun bindFragmentOptions(target: SupportFragment, containerId: Int, record: TransactionRecord) {
            val args = getArguments(target)
            args.putInt(SupportTransaction.FRAGMENTATION_CONTAINER_ID, containerId)
            args.putString(SupportTransaction.FRAGMENTATION_TAG, record.tag)
            args.putString(SupportTransaction.FRAGMENTATION_SIMPLE_NAME, target.javaClass.simpleName)
            args.putString(SupportTransaction.FRAGMENTATION_FULL_NAME, target.javaClass.name)
            args.putBoolean(SupportTransaction.FRAGMENTATION_PLAY_ENTER_ANIM, record.displayEnterAnim)
            args.putBoolean(SupportTransaction.FRAGMENTATION_INIT_LIST, false)
            args.putBoolean(SupportTransaction.FRAGMENTATION_SAVED_INSTANCE, false)
            args.putBoolean(SupportTransaction.FRAGMENTATION_DONT_DISPLAY_SELF_POP_ANIM, record.dontDisplaySelfPopAnim)
        }

        fun setStartForResult(from: SupportFragment?, to: SupportFragment, requestCode: Int) {
            getArguments(from).apply { putInt(SupportTransaction.FRAGMENTATION_REQUEST_CODE, requestCode) }
            getArguments(to).apply { putInt(SupportTransaction.FRAGMENTATION_FROM_REQUEST_CODE, requestCode) }
        }

        fun setResult(target: SupportFragment?, resultCode: Int, data: Bundle) {
            val bundle = getArguments(target)
            bundle.putInt(SupportTransaction.FRAGMENTATION_RESULT_CODE, resultCode)
            bundle.putBundle(SupportTransaction.FRAGMENTATION_RESULT_DATA, data)
        }

        private fun popTo(fm: FragmentManager?, cls: Class<*>?, includeTarget: Boolean, run: Runnable?) {
            fm ?: return
            val remove = fm.getLastFragment() ?: return
            val target = cls?.let { fm.findFragment<SupportFragment>(cls) } ?: return

            val targetIndex = fm.fragments.indexOf(target)
            val removeIndex = fm.fragments.indexOf(remove)

            val ft = fm.beginTransaction()
            ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
            val removeList = fm.fragments.subList(targetIndex, removeIndex)
            if (!includeTarget) {
                removeList.remove(target)
            }
            removeList.filter { it is SupportFragment }.forEach { ft.remove(it) }
            ft.remove(remove)
            ft.supportCommit(run)
        }
    }
}
